ring/util_test.go

```diff
@@ -232,15 +240,123 @@ func TestWaitRingStabilityShouldReturnErrorIfMaxWaitingIsReached(t *testing.T) {
+	elapsedTime := time.Since(startTime)
+
+	assert.GreaterOrEqual(t, elapsedTime.Milliseconds(), maxWaiting.Milliseconds())
+}
+
+func TestWaitRingTokensStability_ShouldReturnErrorIfInstancesAddedAndMaxWaitingIsReached(t *testing.T) {
+	t.Parallel()
+
+	const (
+		minStability = 2 * time.Second
+		maxWaiting   = 7 * time.Second
+	)
+
+	ring := createStartingRing()
+
+	done := addInstancesPeriodically(ring)
+	defer close(done)
+
+	startTime := time.Now()
+	require.Equal(t, context.DeadlineExceeded, WaitRingTokensStability(context.Background(), ring, Reporting, minStability, maxWaiting))
+	elapsedTime := time.Since(startTime)
+
+	assert.GreaterOrEqual(t, elapsedTime.Milliseconds(), maxWaiting.Milliseconds())
+}
+
+// Keep changing the ring in a way to avoid repeating the same set of states for at least 2 sec
+func changeStatePeriodically(ring *Ring) chan struct{} {
+	done := make(chan struct{})
+	go func() {
+		instanceToMutate := "instance-1"
+		states := []InstanceState{PENDING, JOINING, ACTIVE, LEAVING}
+		stateIdx := 0
+
+		for states[stateIdx] != ring.ringDesc.Ingesters[instanceToMutate].State {
+			stateIdx++
+		}
+
+		for {
+			select {
+			case <-done:
+				return
+			case <-time.After(time.Second):
+				stateIdx++
+				ring.mtx.Lock()
+				ringDesc := ring.ringDesc
+				desc := InstanceDesc{Addr: instanceToMutate, State: states[stateIdx%len(states)], Timestamp: time.Now().Unix()}
+				ringDesc.Ingesters[instanceToMutate] = desc
+				ring.mtx.Unlock()
+			}
+		}
+	}()
+
+	return done
+}
+
+func TestWaitRingStability_ShouldReturnErrorIfInstanceStateIsChangingAndMaxWaitingIsReached(t *testing.T) {
+	t.Parallel()
+
+	const (
+		minStability = 2 * time.Second
+		maxWaiting   = 7 * time.Second
+	)
+
+	ring := createStartingRing()
+
+	// Keep changing the ring.
+	done := changeStatePeriodically(ring)
+	defer close(done)
 
 	startTime := time.Now()
 	require.Equal(t, context.DeadlineExceeded, WaitRingStability(context.Background(), ring, Reporting, minStability, maxWaiting))
 	elapsedTime := time.Since(startTime)
 
-	assert.InDelta(t, maxWaiting, elapsedTime, float64(2*time.Second))
+	assert.GreaterOrEqual(t, elapsedTime.Milliseconds(), maxWaiting.Milliseconds())
+}
+
+func TestWaitRingTokensStability_ShouldReturnOnceMinStabilityOfInstancesHasBeenReachedWhileStateCanChange(t *testing.T) {
+	t.Parallel()
+
+	const (
+		minStability = 2 * time.Second
+		maxWaiting   = 7 * time.Second
+	)
+
+	ring := createStartingRing()
+
+	// Keep changing the ring.
+	done := changeStatePeriodically(ring)
+	defer close(done)
+
+	startTime := time.Now()
+	require.Equal(t, nil, WaitRingTokensStability(context.Background(), ring, Reporting, minStability, maxWaiting))
+	elapsedTime := time.Since(startTime)
+
+	assert.GreaterOrEqual(t, elapsedTime.Milliseconds(), minStability.Milliseconds())
+	assert.Less(t, elapsedTime.Milliseconds(), 2*minStability.Milliseconds())
```

Comment:
```suggestion
	assert.Less(t, elapsedTime, 2*minStability)
```

(Everywhere, not just here.)

Reply:
lol , true, don't know why it was like that in the first place, removed